service/src/main/java/org/jboss/aerogear/unifiedpush/service/impl/PushApplicationServiceImpl.java

```diff
@@ -41,10 +41,16 @@ public PushApplicationServiceImpl() {
     }
 
     @Override
-    public void addPushApplication(PushApplication pushApp) {
-
+    public void addPushApplication(PushApplication pushApp) throws IllegalArgumentException {
         pushApp.setDeveloper(loginName.get());
-        pushApplicationDao.create(pushApp);
+
+        final String pushApplicationID = pushApp.getPushApplicationID();
+
+        if (findByPushApplicationID(pushApplicationID) != null) {
+            throw new IllegalArgumentException("PushApplicationID exits already: " + pushApplicationID);
+        } else {
+            pushApplicationDao.create(pushApp);
```

Comment:
I'd guess that the create, deep down below, throws some "ID already exists" error.

So we do not need to query the DB before

Reply:
I think that querying the DB before is more robust. As @ziccardi suggested, the transaction might not be committed right away so the "ID already exists" error wouldn't be caught by the time the client is trying to add the app. By querying the DB prior the create, we can handle this and for instance warn the client that it is performing an illegal transaction.